File: backend/src/main/java/at/ac/tuwien/sepm/groupphase/backend/service/CategoryService.java
package at.ac.tuwien.sepm.groupphase.backend.service;

import at.ac.tuwien.sepm.groupphase.backend.entity.Category;
import org.springframework.data.domain.Page;
import at.ac.tuwien.sepm.groupphase.backend.exception.NotFoundException;


import java.util.Collection;
import java.util.List;
/**
 * Service class handling business logic for Categories.
 * */

public interface CategoryService {
    /**
     * Creates a new category that is eventually associated to a product.
     *
     * @param category the newly added product
     * @return the newly created category
     * @throws RuntimeException occurs during database operations
     */
    Category createCategory(Category category);

    /**
     * Gets all Categories that were previously created in a  paginated Manner.
     *
     * @param page the current page
     * @param pageSize number of entries per page
     * @return all categories that were previously saved in a page Object given by page and pageSize
     * @throws RuntimeException occurs during database operations
     */
    Page<Category> getAllCategoriesPerPage(int page, int pageSize);

    /**
     * Gets all Categories that were previously created.
     *
     * @return all categories that were previously saved
     * @throws RuntimeException occurs during database operations
     */
    List<Category> getAllCategories();

    /**
     * Gets the total number of categories.
     *
     * @return the number of total categories
     * @throws RuntimeException occurs during database operations
     * */
    Long getCategoriesCount();

    /**
     * Updates an already existing category entity in the database with the given Id.
     *
     * @param categoryId the id of the category to execute the update
     * @param category the new category with updated fields
     * @throws RuntimeException upon encountering errors with the database
     */
    void updateCategory(Long categoryId, Category category);

    /**
     * Deletes a category entity in the database with the given Id.
     *
     * @param categoryId the id of the category to execute the delete action
     * @throws RuntimeException upon encountering errors with the database
     */
    void deleteCategory(Long categoryId);

    /**
     * Gets the category entity  with the given Id.
     *
     * @param categoryId the id of the category to retrieve from the database
     * @return the requested category specified by the id
     * @throws NotFoundException is being thrown when the request category is not found
     * @throws RuntimeException is being thrown if some error occurs during data processing
     *
     */
    Category getCategoryById(Long categoryId);



}

File: frontend/src/app/services/category.service.ts
import { Injectable } from '@angular/core';
import {HttpClient, HttpHeaders, HttpParams} from '@angular/common/http';
import {Globals} from '../global/globals';
import {Observable} from 'rxjs';
import {Category} from '../dtos/category';
import {Pagination} from '../dtos/pagination';
import {OperatorAuthService} from './auth/operator-auth.service';

@Injectable({
  providedIn: 'root'
})
export class CategoryService {

  private categoryBaseUri: string = this.globals.backendUri + '/categories';

  constructor(private httpClient: HttpClient, private globals: Globals, private operatorAuthService: OperatorAuthService) {
  }

  /**
   * Loads all categories from the backend
   */
  getCategories(): Observable<Category[]> {
    return this.httpClient.get<Category[]>(this.categoryBaseUri + '/all');
  }

  /**
   * Loads all categories from the backend in a paginated manner specified by page and pageSize
   *
   * @param page the current page the user is in
   * @param pageCount number of all entries in a page
   *
   * @return An Observable with paginated category
   */
  getCategoriesPerPage(page: number, pageCount): Observable<Pagination<Category>> {
    const params = new HttpParams()
      .set(this.globals.requestParamKeys.pagination.page, String(page))
      .set(this.globals.requestParamKeys.pagination.pageCount, String(pageCount));

    return this.httpClient.get<Pagination<Category>>(this.categoryBaseUri, {
      params,
      headers: this.getHeadersForOperator()
    });
  }

  /**
   * adds a new category to the backend
   *
   * @param category the new category to be added
   * @return An Observable with the newly added category
   */
  addCategory(category: Category): Observable<Category> {
    return this.httpClient.post<Category>(this.categoryBaseUri, category, {
      headers: this.getHeadersForOperator()
    });
  }

  /**
   * updates an already existing category to the backend
   *
   * @param categoryId the Id of the category to execute the update
   * @param category the updated category
   * @return An Observable with no return value
   */
  updateCategory(categoryId: number, category: Category): Observable<void> {
    return this.httpClient.put<void>(this.categoryBaseUri + '/' + categoryId, category, {
      headers: this.getHeadersForOperator()
    });
  }
  /**
   * deletes a category in the backend
   *
   * @param categoryId the Id of the category to execute the delete action
   * @return An Observable with no return value
   */
  deleteCategory(categoryId: number): Observable<void> {
    return this.httpClient.delete<void>(this.categoryBaseUri + '/' + categoryId, {
      headers: this.getHeadersForOperator()
    });
  }
  /**
   * Gets a category specified by the id
   *
   * @param categoryId the Id of the category to retrieve from the database
   * @return An observable with the requested category
   */
  getCategoryById(categoryId: number) {
    return this.httpClient.get<Category>(this.categoryBaseUri + '/' + categoryId, {
      headers: this.getHeadersForOperator()
    });
  }

  private getHeadersForOperator(): HttpHeaders {
    return new HttpHeaders()
      .set('Authorization', `Bearer ${this.operatorAuthService.getToken()}`);
  }
}
